File: C++/source/Framework/Messaging/LightHouse/LightPulseDispatchWorker.cpp
/***************************************************************************
 *   Copyright (C) 20016 Vidyasagara Guntaka                               *
 *   All rights reserved.                                                  *
 *   Author : Vidyasagara Reddy Guntaka                                    *
 ***************************************************************************/

#include "Framework/Messaging/LightHouse/LightPulseDispatchWorker.h"

#include "Framework/Core/WaveFrameworkMessages.h"
#include "Framework/Messaging/LightHouse/LightPulse.h"
#include "Framework/Messaging/LightHouse/LightPulseDispatchMessage.h"
#include "Framework/Messaging/LightHouse/LightPulseDispatchObjectManager.h"
#include "Framework/Messaging/LightHouse/LightPulseDispatchTypes.h"
#include "Framework/Messaging/LightHouse/LightPulseRegistrationRepository.h"
#include "Framework/Messaging/LightHouse/Test/LightHouseTestLightPulse1.h"
#include "Framework/Messaging/MessageFactory/WaveMessageFactory.h"
#include "Framework/Utils/FrameworkToolKit.h"

using WaveNs::LightPulseDispatchObjectManager;

namespace WaveNs
{

LightPulseDispatchWorker::LightPulseDispatchWorker (LightPulseDispatchObjectManager *pLightPulseDispatchObjectManager)
    : WaveWorker (pLightPulseDispatchObjectManager)
{
    addOperationMap (LIGHT_PULSE_DISPATCH_RECEIVED_LIGHT_PULSE, reinterpret_cast<WaveMessageHandler> (&LightPulseDispatchWorker::dispatchReceivedLightPulseMessageHandler));
}

LightPulseDispatchWorker::~LightPulseDispatchWorker ()
{
}

void LightPulseDispatchWorker::dispatchReceivedLightPulseMessageHandler (LightPulseDispatchMessage *pLightPulseDispatchMessage)
{
    trace (TRACE_LEVEL_DEVEL, "LightPulseDispatchWorker::dispatchReceivedLightPulseMessageHandler : Entering ...");

    waveAssert (NULL != pLightPulseDispatchMessage, __FILE__, __LINE__);

    const string seralizedLightPulseString = pLightPulseDispatchMessage->getLightPulseSerializedString ();

    trace (TRACE_LEVEL_DEVEL, "LightPulseDispatchWorker::dispatchReceivedLightPulseMessageHandler : Dispatching a Light Pulse ...\r\n" + seralizedLightPulseString);

    const string lightPulseName = LightPulse::getLightPulseName (seralizedLightPulseString.c_str ());

    trace (TRACE_LEVEL_DEVEL, "LightPulseDispatchWorker::dispatchReceivedLightPulseMessageHandler : Light Pulse Name : " + lightPulseName);

    LightPulse *pLightPulse = WaveMessageFactory::getLightPulseInstance (lightPulseName);

    if (NULL == pLightPulse)
    {
        trace (TRACE_LEVEL_FATAL, "LightPulseDispatchWorker::dispatchReceivedLightPulseMessageHandler : Please have the corresponding Object Manager implement  createLightPulseInstance for : " + lightPulseName);
        waveAssert (false, __FILE__, __LINE__);
    }

    pLightPulse->loadFromSerializedData2 (seralizedLightPulseString);

    pLightPulse->setSenderIpAddressFromClientPerspective (pLightPulseDispatchMessage->getSenderIpAddress ());

    LightPulseRegistrationRepository *pLightPulseRegistrationRepository = LightPulseRegistrationRepository::getInstance ();

    waveAssert (NULL != pLightPulseRegistrationRepository, __FILE__, __LINE__);

    set<WaveServiceId> servicesRegisteredForLightPulse = pLightPulseRegistrationRepository->getLightPulseRecepientsForLightPulseName (lightPulseName);

    set<WaveServiceId>::iterator element    = servicesRegisteredForLightPulse.begin ();
    set<WaveServiceId>::iterator endElement = servicesRegisteredForLightPulse.end   ();

    while (endElement != element)
    {
        WaveServiceId waveServiceId = *element;

        WaveDeliverLightPulseMessage *pWaveDeliverLightPulseMessage = new WaveDeliverLightPulseMessage (waveServiceId);

        pWaveDeliverLightPulseMessage->setPLightPulse (pLightPulse);

        WaveMessageStatus sendStatus = sendOneWay (pWaveDeliverLightPulseMessage);

        if (WAVE_MESSAGE_SUCCESS != sendStatus)
        {
            tracePrintf (TRACE_LEVEL_ERROR, true, false, "LightPulseDispatchWorker::dispatchReceivedLightPulseMessageHandler Could not deliver Light Pulse of type %s to service : %s, status : %s", lightPulseName.c_str (), (FrameworkToolKit::getServiceNameById (waveServiceId)).c_str (), (FrameworkToolKit::localize (sendStatus)).c_str ());
        }
        else
        {
            tracePrintf (TRACE_LEVEL_DEVEL, true, false, "LightPulseDispatchWorker::dispatchReceivedLightPulseMessageHandler : Trying to deliver Light Pulse of type %s to service : %s", lightPulseName.c_str (), (FrameworkToolKit::getServiceNameById (waveServiceId)).c_str ());
        }

        element++;
    }

    pLightPulseDispatchMessage->setCompletionStatus (WAVE_MESSAGE_SUCCESS);
    reply (pLightPulseDispatchMessage);
}

}

File: Java/source/com/CxWave/Wave/Framework/Messaging/LightHouse/LightPulseDispatchWorker.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka & CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.Framework.Messaging.LightHouse;

import java.util.Set;

import com.CxWave.Wave.Framework.Core.Messages.WaveDeliverLightPulseMessage;
import com.CxWave.Wave.Framework.Messaging.MessageFactory.WaveMessageFactory;
import com.CxWave.Wave.Framework.ObjectModel.WaveWorker;
import com.CxWave.Wave.Framework.ObjectModel.WaveWorkerPriority;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.Cardinality;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.OwnerOM;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.WorkerPriority;
import com.CxWave.Wave.Framework.ToolKits.Framework.FrameworkToolKit;
import com.CxWave.Wave.Framework.Type.WaveServiceId;
import com.CxWave.Wave.Resources.ResourceEnums.ResourceId;
import com.CxWave.Wave.Resources.ResourceEnums.TraceLevel;
import com.CxWave.Wave.Resources.ResourceEnums.WaveMessageStatus;

@OwnerOM (om = LightPulseDispatchObjectManager.class)
@Cardinality (1)
@WorkerPriority (WaveWorkerPriority.WAVE_WORKER_PRIORITY_0)
public class LightPulseDispatchWorker extends WaveWorker
{
    public LightPulseDispatchWorker (final LightPulseDispatchObjectManager lightPulseDispatchObjectManager)
    {
        super (lightPulseDispatchObjectManager);
    }

    private void dispatchReceivedLightPulseMessageHandler (final LightPulseDispatchMessage lightPulseDispatchMessage)
    {
        infoTracePrintf ("LightPulseDispatchWorker.dispatchReceivedLightPulseMessageHandler : Receied a light Pulse : %s", lightPulseDispatchMessage.getLightPulseSerializedString ());

        waveAssert (null != lightPulseDispatchMessage);

        final String seralizedLightPulseString = lightPulseDispatchMessage.getLightPulseSerializedString ();

        trace (TraceLevel.TRACE_LEVEL_DEVEL, "LightPulseDispatchWorker.dispatchReceivedLightPulseMessageHandler : Dispatching a Light Pulse ...\r\n" + seralizedLightPulseString);

        final String lightPulseName = LightPulse.getLightPulseName (seralizedLightPulseString);

        trace (TraceLevel.TRACE_LEVEL_DEVEL, "LightPulseDispatchWorker.dispatchReceivedLightPulseMessageHandler : Light Pulse Name : " + lightPulseName);

        final LightPulse lightPulse = WaveMessageFactory.getLightPulseInstance (lightPulseName);

        if (null == lightPulse)
        {
            trace (TraceLevel.TRACE_LEVEL_FATAL, "LightPulseDispatchWorker.dispatchReceivedLightPulseMessageHandler : Please have the corresponding Object Manager implement  createLightPulseInstance for : " + lightPulseName);
            waveAssert (false);
        }

        lightPulse.loadFromSerializedData2 (seralizedLightPulseString);

        final LightPulseRegistrationRepository lightPulseRegistrationRepository = LightPulseRegistrationRepository.getInstance ();

        waveAssert (null != lightPulseRegistrationRepository);

        final Set<WaveServiceId> servicesRegisteredForLightPulse = lightPulseRegistrationRepository.getLightPulseRecepientsForLightPulseName (lightPulseName);

        for (final WaveServiceId waveServiceId : servicesRegisteredForLightPulse)
        {
            final WaveDeliverLightPulseMessage waveDeliverLightPulseMessage = new WaveDeliverLightPulseMessage (waveServiceId);

            waveDeliverLightPulseMessage.setLightPulse (lightPulse);

            final WaveMessageStatus sendStatus = sendOneWay (waveDeliverLightPulseMessage);

            if (WaveMessageStatus.WAVE_MESSAGE_SUCCESS != sendStatus)
            {
                tracePrintf (TraceLevel.TRACE_LEVEL_ERROR, true, false, "LightPulseDispatchWorker.dispatchReceivedLightPulseMessageHandler Could not deliver Light Pulse of type %s to service : %s, status : %s", lightPulseName, FrameworkToolKit.getServiceNameById (waveServiceId), FrameworkToolKit.localize (sendStatus));
            }
            else
            {
                tracePrintf (TraceLevel.TRACE_LEVEL_DEVEL, true, false, "LightPulseDispatchWorker.dispatchReceivedLightPulseMessageHandler : Trying to deliver Light Pulse of type %s to service : %s", lightPulseName, FrameworkToolKit.getServiceNameById (waveServiceId));
            }
        }

        lightPulseDispatchMessage.setCompletionStatus (ResourceId.WAVE_MESSAGE_SUCCESS);
        reply (lightPulseDispatchMessage);
    }
}
